src/Ouroboros/Storage/Immutable/FS.hs

```diff
@@ -0,0 +1,900 @@
+                                if n == 2 then (acc <> " " <> BS.singleton w8, 1)
+                                          else (acc <> BS.singleton w8, n + 1)
+                            ) (mempty, 0 :: Int)
+
+
+drawTree :: FsTree a String -> String
+drawTree  = unlines . draw
+  where
+    draw :: FsTree a String -> [String]
+    draw (FileOnDisk f _)     = ["(f) " <> f]
+    draw (FolderOnDisk x ts0) = lines x ++ drawSubTrees ts0
+      where
+        drawSubTrees [] = []
+        drawSubTrees [t] =
+            "|" : shiftT "`- " "   " (draw t)
+        drawSubTrees (t:ts) =
+            "|" : shiftT "+- " "|  " (draw t) ++ drawSubTrees ts
+
+        shiftT first other = zipWith (++) (first : repeat other)
+
+-- | Given a 'FilePath' returns a new 'FsTree' populated with all the
+-- directories listed in the input 'FilePath'.
+fsFromFilePath :: HasCallStack => FilePath -> FsTree a String
+fsFromFilePath fp =
+    case splitDirectories fp of
+        [] -> error "fsFromFilePath: empty input path."
+        ps -> mockCreateDirectoryIfMissingInternal True ps (FolderOnDisk (head ps) [])
+
+demoMockFS :: MockFS
+demoMockFS = MockFS {
+    getMockFS = FolderOnDisk "/" [
+      FolderOnDisk "usr" [
+                   FolderOnDisk "local" [FolderOnDisk "bin" mempty]
+                 ]
+    , FolderOnDisk "var" [
+                   FolderOnDisk "log"  mempty
+                 , FolderOnDisk "tmp"  [FileOnDisk "foo.txt" mempty]
+                 , FolderOnDisk "mail" mempty
+                 , FolderOnDisk "run"  mempty
+                 ]
+    ]
+  }
+
+newEmptyMockFS :: MockFS
+newEmptyMockFS = MockFS {
+    getMockFS = FolderOnDisk "/" []
+  }
+
+-- Like 'newEmptyMockFS', but allows the user to supply an input 'FilePath' to
+-- create an initial directory on disk.
+newMockFsWith :: FilePath -> MockFS
+newMockFsWith dir = newEmptyMockFS { getMockFS = fsFromFilePath dir }
+
+data SimFsAction m =
+    SimOpen                     FsPath IOMode
+  | SimClose                    (MockHandle m)
+  | SimSeek                     FsPath SeekMode DiskOffset
+  | SimGet                      FsPath Int
+  | SimPut                      FsPath ByteString
+  | SimTruncate                 FsPath Word64
+  | SimListDirectory            FsPath [String]
+  | SimCreateDirectory          FsPath
+  | SimCreateDirectoryIfMissing FsPath Bool
+  deriving Show
+
+deriving instance Eq (SimFsAction IO)
+
+type SimFsEnv m = (FsPath, TVar m MockFS , TVar m [SimFsAction m])
+
+newtype SimFS m a =
+    SimFS { unSimFs :: ReaderT (SimFsEnv m) m a }
+    deriving (Functor, Applicative, Monad)
+
+instance MonadTrans SimFS where
+    lift = SimFS . lift
+
+instance MonadThrow m  => MonadThrow (SimFS m) where
+    throwM = SimFS . lift . throwM
+
+instance MonadCatch m  => MonadCatch (SimFS m) where
+    catch (SimFS ma) handler =
+        SimFS $ catch ma (\e -> let (SimFS r) = handler e in r)
+
+deriving instance MonadMask m => MonadMask (SimFS m)
```

Comment:
Is there a reason for a standalone deriving here, rather than than just add it to the `deriving` clause?